autopilot/contractor/contractor.go

```diff
@@ -189,714 +191,271 @@ func (c *Contractor) Close() error {
+	if err != nil {
+		log.Errorw("adding refreshed contract failed", zap.Error(err), "hk", hk, "fcid", fcid)
+		return api.ContractMetadata{}, false, err
 	}
 
-	// cap the amount of contracts we want to keep to the configured amount
-	for _, contract := range updatedSet {
-		if _, exists := contractData[contract.ID]; !exists {
-			c.logger.Errorf("contract %v not found in contractData", contract.ID)
-		}
-	}
-	if len(updatedSet) > int(ctx.WantedContracts()) {
-		// sort by contract size
-		sort.Slice(updatedSet, func(i, j int) bool {
-			return contractData[updatedSet[i].ID] > contractData[updatedSet[j].ID]
-		})
-		for _, contract := range updatedSet[ctx.WantedContracts():] {
-			toStopUsing[contract.ID] = "truncated"
-		}
-		updatedSet = updatedSet[:ctx.WantedContracts()]
-	}
+	// add to renewed set
+	newCollateral := resp.Contract.Revision.MissedHostPayout().Sub(resp.ContractPrice)
+	log.Infow("refresh succeeded",
+		"fcid", refreshedContract.ID,
+		"renewedFrom", contract.ID,
+		"renterFunds", renterFunds.String(),
+		"minNewCollateral", minNewCollateral.String(),
+		"newCollateral", newCollateral.String(),
+	)
+	return refreshedContract, true, nil
+}
 
-	// convert to set of file contract ids
-	var newSet []types.FileContractID
-	for _, contract := range updatedSet {
-		newSet = append(newSet, contract.ID)
+func (c *Contractor) renewContract(ctx *mCtx, w Worker, contract api.Contract, host api.Host, budget *types.Currency) (cm api.ContractMetadata, proceed bool, err error) {
+	if contract.Revision == nil {
+		return api.ContractMetadata{}, true, errors.New("can't renew contract without a revision")
 	}
+	log := c.logger.With("to_renew", contract.ID, "hk", contract.HostKey, "hostVersion", host.Settings.Version, "hostRelease", host.Settings.Release)
+
+	// convenience variables
+	settings := host.Settings
+	pt := host.PriceTable.HostPriceTable
+	fcid := contract.ID
+	rev := contract.Revision
+	hk := contract.HostKey
 
-	// update contract set
-	err = c.bus.SetContractSet(ctx, ctx.ContractSet(), newSet)
+	// fetch consensus state
+	cs, err := c.bus.ConsensusState(ctx)
 	if err != nil {
-		return false, err
+		return api.ContractMetadata{}, false, err
 	}
 
-	// return whether the maintenance changed the contract set
-	return c.computeContractSetChanged(mCtx, currentSet, updatedSet, formed, refreshed, renewed, toStopUsing, contractData), nil
-}
-
-func (c *Contractor) computeContractSetChanged(ctx *mCtx, oldSet, newSet, formed []api.ContractMetadata, refreshed, renewed []renewal, toStopUsing map[types.FileContractID]string, contractData map[types.FileContractID]uint64) bool {
-	name := ctx.ContractSet()
+	// calculate the renter funds for the renewal a.k.a. the funds the renter will
+	// be able to spend
+	minRenterFunds, _ := initialContractFundingMinMax(ctx.AutopilotConfig())
+	renterFunds := renewFundingEstimate(minRenterFunds, contract.TotalCost, contract.RenterFunds(), log)
```

Comment:
That is true but then unit testing becomes more annoying since we need to construct a full contract to check the 2 fields we are interested in. That's why I opted for minimal obsolete fields as a tradeoff on those lower level functions.